File: migration/run_demo.ts
#!/usr/bin/env npx ts-node
/**
 * News Pipeline Demo
 *
 * This script demonstrates the full pipeline:
 * 1. Fetches news from GDELT and RSS sources
 * 2. Parses and normalizes articles
 * 3. Filters and deduplicates
 * 4. Outputs to JSONL file
 *
 * Usage:
 *   npx ts-node run_demo.ts
 *   # or
 *   npm run demo
 */

import { NewsPipeline, createJsonlStorage, logger } from './src';
import type { PipelineResult } from './src/types';

// Enable debug logging
process.env.DEBUG = 'true';

async function main(): Promise<void> {
  console.log('='.repeat(60));
  console.log('Trading News Scanner - Demo');
  console.log('='.repeat(60));
  console.log();

  // Create storage adapter
  const storage = createJsonlStorage('./migration/output');

  // Create pipeline with configuration
  const pipeline = new NewsPipeline({
    // Fetch from both GDELT and RSS
    useGdelt: true,
    useRss: true,
    useIntel: true,

    // Categories to fetch
    categories: ['finance', 'tech', 'politics'],

    // Optional filter: only trading-relevant topics
    filter: {
      // topics: ['FINANCE', 'CRYPTO', 'TECH'],
      // max_age_ms: 24 * 60 * 60 * 1000, // Last 24 hours
    },

    // Use CORS proxies for RSS (needed in browser, optional in Node)
    corsProxies: [
      'https://api.allorigins.win/raw?url=',
    ],

    // Output storage
    storage,
  });

  console.log('Starting pipeline...');
  console.log();

  try {
    const result: PipelineResult = await pipeline.run();

    // Print summary
    console.log();
    console.log('='.repeat(60));
    console.log('Pipeline Results');
    console.log('='.repeat(60));
    console.log();
    console.log(`Fetched:      ${result.stats.fetched} items`);
    console.log(`Parsed:       ${result.stats.parsed} items`);
    console.log(`Filtered:     ${result.stats.filtered} items removed`);
    console.log(`Deduplicated: ${result.stats.deduplicated} duplicates removed`);
    console.log(`Final:        ${result.items.length} items`);
    console.log(`Duration:     ${result.stats.duration_ms}ms`);
    console.log();

    if (result.errors.length > 0) {
      console.log('Errors:');
      for (const error of result.errors) {
        console.log(`  - [${error.stage}] ${error.source || ''}: ${error.message}`);
      }
      console.log();
    }

    // Print sample items
    console.log('Sample Items (first 5):');
    console.log('-'.repeat(60));

    for (const item of result.items.slice(0, 5)) {
      console.log();
      console.log(`Title:     ${item.title.slice(0, 70)}${item.title.length > 70 ? '...' : ''}`);
      console.log(`Source:    ${item.source}`);
      console.log(`Published: ${item.published_at}`);
      console.log(`Topics:    ${item.topics.join(', ') || 'none'}`);
      console.log(`Tickers:   ${item.tickers.join(', ') || 'none'}`);
      console.log(`Alert:     ${item.metadata.is_alert ? `YES (${item.metadata.alert_keyword})` : 'no'}`);
      console.log(`URL:       ${item.url}`);
    }

    console.log();
    console.log('-'.repeat(60));
    console.log(`Output saved to: ${storage.getFilePath()}`);
    console.log();

    // Print alert items if any
    const alerts = result.items.filter((item) => item.metadata.is_alert);
    if (alerts.length > 0) {
      console.log('='.repeat(60));
      console.log(`ALERTS (${alerts.length} items with alert keywords)`);
      console.log('='.repeat(60));

      for (const alert of alerts.slice(0, 10)) {
        console.log(`[${alert.metadata.alert_keyword?.toUpperCase()}] ${alert.title.slice(0, 60)}`);
      }

      if (alerts.length > 10) {
        console.log(`... and ${alerts.length - 10} more alerts`);
      }
      console.log();
    }

    // Print topic distribution
    const topicCounts = new Map<string, number>();
    for (const item of result.items) {
      for (const topic of item.topics) {
        topicCounts.set(topic, (topicCounts.get(topic) || 0) + 1);
      }
    }

    if (topicCounts.size > 0) {
      console.log('Topic Distribution:');
      const sorted = Array.from(topicCounts.entries()).sort((a, b) => b[1] - a[1]);
      for (const [topic, count] of sorted) {
        const bar = '#'.repeat(Math.min(count, 30));
        console.log(`  ${topic.padEnd(12)} ${count.toString().padStart(3)} ${bar}`);
      }
      console.log();
    }

    // Print source distribution
    const sourceCounts = new Map<string, number>();
    for (const item of result.items) {
      sourceCounts.set(item.source, (sourceCounts.get(item.source) || 0) + 1);
    }

    console.log('Source Distribution (top 10):');
    const sortedSources = Array.from(sourceCounts.entries())
      .sort((a, b) => b[1] - a[1])
      .slice(0, 10);

    for (const [source, count] of sortedSources) {
      console.log(`  ${source.padEnd(25)} ${count}`);
    }
    console.log();

  } catch (error) {
    console.error('Pipeline failed:', error);
    process.exit(1);
  }
}

// Run the demo
main().catch(console.error);

File: py_migration/run_demo.py
#!/usr/bin/env python3
"""
News Pipeline Demo

Demonstrates the full pipeline:
1. Fetches news from GDELT and RSS sources
2. Parses and normalizes articles
3. Filters and deduplicates
4. Outputs to JSONL file

Usage:
    python run_demo.py
    python run_demo.py --categories finance tech
    python run_demo.py --alerts-only
"""

import asyncio
import argparse
from collections import Counter

from news_scanner import (
    NewsPipeline,
    PipelineResult,
    JsonlStorage,
)
from news_scanner.pipeline import PipelineOptions, FilterConfig
from news_scanner.utils import logger


def print_results(result: PipelineResult, storage_path: str):
    """Print pipeline results."""
    print()
    print("=" * 60)
    print("Pipeline Results")
    print("=" * 60)
    print()
    print(f"Fetched:      {result.stats.fetched} items")
    print(f"Parsed:       {result.stats.parsed} items")
    print(f"Filtered:     {result.stats.filtered} items removed")
    print(f"Deduplicated: {result.stats.deduplicated} duplicates removed")
    print(f"Stored:       {result.stats.stored} items")
    print(f"Duration:     {result.stats.duration_ms}ms")
    print()

    if result.errors:
        print("Errors:")
        for error in result.errors:
            print(f"  - [{error.stage}] {error.source or ''}: {error.message}")
        print()

    # Sample items
    print("Sample Items (first 5):")
    print("-" * 60)

    for item in result.items[:5]:
        print()
        title = item.title[:70] + "..." if len(item.title) > 70 else item.title
        print(f"Title:     {title}")
        print(f"Source:    {item.source}")
        print(f"Published: {item.published_at}")
        print(f"Topics:    {', '.join(item.topics) or 'none'}")
        print(f"Tickers:   {', '.join(item.tickers) or 'none'}")
        alert_str = f"YES ({item.metadata.alert_keyword})" if item.metadata.is_alert else "no"
        print(f"Alert:     {alert_str}")

    print()
    print("-" * 60)
    print(f"Output saved to: {storage_path}")
    print()

    # Alerts
    alerts = [item for item in result.items if item.metadata.is_alert]
    if alerts:
        print("=" * 60)
        print(f"ALERTS ({len(alerts)} items)")
        print("=" * 60)
        for alert in alerts[:10]:
            kw = alert.metadata.alert_keyword.upper() if alert.metadata.alert_keyword else "?"
            print(f"[{kw}] {alert.title[:55]}...")
        if len(alerts) > 10:
            print(f"... and {len(alerts) - 10} more alerts")
        print()

    # Topic distribution
    topic_counts = Counter()
    for item in result.items:
        for topic in item.topics:
            topic_counts[topic] += 1

    if topic_counts:
        print("Topic Distribution:")
        for topic, count in topic_counts.most_common():
            bar = "#" * min(count, 30)
            print(f"  {topic:<12} {count:>3} {bar}")
        print()

    # Source distribution
    source_counts = Counter(item.source for item in result.items)
    print("Source Distribution (top 10):")
    for source, count in source_counts.most_common(10):
        print(f"  {source:<25} {count}")
    print()


async def main():
    parser = argparse.ArgumentParser(description="News Pipeline Demo")
    parser.add_argument(
        "--categories",
        nargs="+",
        default=["finance", "tech", "politics"],
        help="Categories to fetch"
    )
    parser.add_argument(
        "--no-gdelt",
        action="store_true",
        help="Skip GDELT source"
    )
    parser.add_argument(
        "--no-rss",
        action="store_true",
        help="Skip RSS sources"
    )
    parser.add_argument(
        "--no-intel",
        action="store_true",
        help="Skip intel sources"
    )
    parser.add_argument(
        "--alerts-only",
        action="store_true",
        help="Only show alert items"
    )
    parser.add_argument(
        "--max-age",
        type=int,
        help="Max age in hours"
    )
    parser.add_argument(
        "--output",
        default="./output",
        help="Output directory"
    )
    parser.add_argument(
        "--debug",
        action="store_true",
        help="Enable debug logging"
    )

    args = parser.parse_args()

    if args.debug:
        logger.enable_debug()

    print("=" * 60)
    print("Trading News Scanner - Demo")
    print("=" * 60)
    print()
    print(f"Categories: {', '.join(args.categories)}")
    print(f"Sources: GDELT={not args.no_gdelt}, RSS={not args.no_rss}, Intel={not args.no_intel}")
    print()

    # Build filter config
    filter_config = None
    if args.alerts_only or args.max_age:
        filter_config = FilterConfig(
            alerts_only=args.alerts_only,
            max_age_hours=args.max_age
        )

    # Create storage
    storage = JsonlStorage.with_timestamp(args.output)

    # Create pipeline
    options = PipelineOptions(
        use_gdelt=not args.no_gdelt,
        use_rss=not args.no_rss,
        use_intel=not args.no_intel,
        categories=args.categories,
        filter_config=filter_config,
        storage=storage
    )

    pipeline = NewsPipeline(options)

    print("Starting pipeline...")
    result = await pipeline.run()

    print_results(result, storage.get_path())
    print("Done!")


if __name__ == "__main__":
    asyncio.run(main())
